lib/calendar_interval.ex

```diff
@@ -32,6 +33,29 @@ defmodule CalendarInterval do
 
   @microsecond {:microsecond, 6}
 
+  @doc """
+  Calendar callback that adds years and months to a naive datetime.
+
+  `step` can be positive or negative.
+
+  When streaming intervals, this is the callback that increments years
+  and months. Incrementing other time precision is managed directly through
+  `NaiveDateTime.add/3`.
+  """
+  @callback add(Calendar.naive_datetime(), precision(), step :: integer()) ::
+              {Calendar.year(), Calendar.month(), Calendar.day()}
```

Comment:
Yes adding to dates is tricky, however:

1. You're already doing it for the `Calendar.ISO` calendar - required of course to support `next/2` and `prev/2`. All this PR does is formalise the notion of adding a time period to a date as a `@callback`. 

2. And as you also note, the semantics of an interval with a precision actually avoids the issue you mention.  Its one of the beautiful things about intervals and why I look forward to getting rid of my own poor interval implementation and adopting yours.

Currently this PR has to have a different implementation because `Calendar.ISO` doesn't have the callbacks to supporting adding the necessary precision to an interval.  Each of the calendar structs has a form of `add/2` (`NaiveDateTime`, `Date` and `Time` but the semantics are different (days for `Date`, seconds for the others.  It would nice to harmonise this as a standard calendar callback.  But it doesn't have to be done - and therefore could be deferred as a discussion until later.  I certainly plan to declare the `CalendarInterval` `@behaviour` in my own calendars if you accept this PR.

Reply:
> It would nice to harmonise this as a standard calendar callback.

right, my understanding is modules implementing `@behaviour CalendarInterval` may need this callback, so that's ok. Since you mentioned stdlib, I was just curious what would it be used for by modules implementing `@behaviour Calendar`.

> But it doesn't have to be done - and therefore could be deferred as a discussion until later.

I think keeping it as is is ok but deferring does sound good too. Your call!